Dipping toes into linker friendliness

davidfowl opened the pull request:
Experimenting with linker friendliness in ASP.NET Core. This investigation did lead to a single API change around the new overload of UseStartup that takes a func (it was changed a generic).

- Annotate a couple of methods that do registration specifying a type or T 
 - UseMiddleware
 - UseStartup
 - UseHub
 - MapHub

Here's an example of what this looks like today in aggressive mode https://github.com/rynowak/link-a-thon/blob/master/src/ApiTemplate/Linker.xml

davidfowl committed:
Dipping toes into linkerfriendliness

JamesNK commented on src/Hosting/Hosting/src/WebHostBuilderExtensions.cs:
```diff
@@ -67,7 +71,7 @@ private static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Actio
         /// <param name="startupFactory">A delegate that specifies a factory for the startup class.</param>
         /// <returns>The <see cref="IWebHostBuilder"/>.</returns>
-        public static IWebHostBuilder UseStartup(this IWebHostBuilder hostBuilder, Func<WebHostBuilderContext, object> startupFactory)
+        public static IWebHostBuilder UseStartup<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]TStartup>(this IWebHostBuilder hostBuilder, Func<WebHostBuilderContext, TStartup> startupFactory)
```
Huh, I didn't know you use attributes like this. Is it new?

davidfowl replied:
Yes https://github.com/dotnet/designs/blob/master/accepted/2020/linking-libraries.md

javiercn replied:
Thank god this is new, I saw it on twitter and I was starting to feel dumb after 10+ years not knowing you could do this :)

JamesNK commented on src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs:
```diff
@@ -213,7 +214,7 @@ public IWebHostBuilder UseStartup(Type startupType)
         }
 
-        public IWebHostBuilder UseStartup(Func<WebHostBuilderContext, object> startupFactory)
+        public IWebHostBuilder UseStartup<TStartup>(Func<WebHostBuilderContext, TStartup> startupFactory)
```
Dumb question but why is this better? Wouldn't the startup type be referenced in the func when it is created?

JamesNK replied:
I think I know. It's because we want to ensure Configure/ConfigureServices as included. Creating the type in the func will ensure the constructor is available, but the methods will get trimmed because they're not used.

A generic type will hit the method with an attribute on the generic type arg that ensures methods are preserved.

davidfowl replied:
Yep

JamesNK commented on src/SignalR/server/Core/src/SignalRConnectionBuilderExtensions.cs:
```diff
@@ -13,13 +14,15 @@ namespace Microsoft.AspNetCore.SignalR
         /// <param name="connectionBuilder">The connection to configure.</param>
         /// <returns>The same instance of the <see cref="IConnectionBuilder"/> for chaining.</returns>
-        public static IConnectionBuilder UseHub<THub>(this IConnectionBuilder connectionBuilder) where THub : Hub
+        public static IConnectionBuilder UseHub<[DynamicallyAccessedMembers(HubAccessibility)]THub>(this IConnectionBuilder connectionBuilder) where THub : Hub
```
Hmm, this should be done in gRPC.

davidfowl replied:
Yes

davidfowl replied:
https://github.com/rynowak/link-a-thon/blob/master/src/GrpcService/Linker.xml

JamesNK reviewed (commented):
To properly certify asp.net core as linker friendly I think we need functional tests that run with linking enabled. This seems like an easy thing to unknowingly regress.

davidfowl commented:
> I think to properly certify asp.net core as linker friendly I think we need functional tests that run with linking enabled. This seems like an easy thing to unknowingly regress.

2 things:
- I agree in the fullness of time we'll need ways to test this so we don't regress behavior
- Aggressive mode isn't a default linker option and it's non-trivial to turn on so the stakes are super low. 

This doesn't change any defaults but it allows us to make progress in understanding how to make the framework more linkable.

davidfowl committed:
Fixed tests

davidfowl commented:
UseMiddleware is the most critical because we use it on our own components. The rest is user code which we would probably avoid trimming by default.

JamesNK commented on src/Hosting/Hosting/src/WebHostBuilderExtensions.cs:
```diff
@@ -151,7 +155,7 @@ public static IWebHostBuilder UseStartup(this IWebHostBuilder hostBuilder, Type
         /// <typeparam name ="TStartup">The type containing the startup methods for the application.</typeparam>
         /// <returns>The <see cref="IWebHostBuilder"/>.</returns>
-        public static IWebHostBuilder UseStartup<TStartup>(this IWebHostBuilder hostBuilder) where TStartup : class
+        public static IWebHostBuilder UseStartup<[DynamicallyAccessedMembers(StartupAccessibility)]TStartup>(this IWebHostBuilder hostBuilder) where TStartup : class
```
nit: is there any guidance or best practice around generic attributes and overloads. It looks like this method calls the earlier UseStartup which will have the attribute so it probably isn't necessary. Not a big deal either way.

davidfowl replied:
I think I'm supposed to apply this attribute transitively through the codebase or the linker will warn.

JamesNK commented on src/Hosting/Hosting/test/WebHostBuilderTests.cs:
```diff
@@ -96,7 +96,7 @@ public async Task MultipleUseStartupCallsLastWins(IWebHostBuilder builder)
             var host = builder.UseServer(server)
                               .UseStartup<StartupCtorThrows>()
-                              .UseStartup(context => throw new InvalidOperationException("This doesn't run"))
+                              .UseStartup<object>(context => throw new InvalidOperationException("This doesn't run"))
```
I'm guessing we're ok with this slight usability regression. The generic arg will always need to be explicitly specified because it can't be inferred from the return type.

davidfowl replied:
That's right. I don't see this being common.

JamesNK commented on src/Hosting/Hosting/src/WebHostBuilderExtensions.cs:
```diff
@@ -67,7 +71,7 @@ private static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Actio
         /// <param name="startupFactory">A delegate that specifies a factory for the startup class.</param>
         /// <returns>The <see cref="IWebHostBuilder"/>.</returns>
-        public static IWebHostBuilder UseStartup(this IWebHostBuilder hostBuilder, Func<WebHostBuilderContext, object> startupFactory)
+        public static IWebHostBuilder UseStartup<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]TStartup>(this IWebHostBuilder hostBuilder, Func<WebHostBuilderContext, TStartup> startupFactory)
```
What happens if someone does this:

`host.UseStartup<object>(c => return MyComplexType());`

We'll be preserving public methods on object, not MyComplexType. Should we validate the return type from the func is the same as the specified type?

davidfowl replied:
Yes this is a tricky one and I agree.

davidfowl committed:
Mark everything transitively

eerhardt commented on src/Hosting/Hosting/src/WebHostBuilderExtensions.cs:
```diff
@@ -67,7 +71,7 @@ private static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Actio
         /// <param name="startupFactory">A delegate that specifies a factory for the startup class.</param>
         /// <returns>The <see cref="IWebHostBuilder"/>.</returns>
-        public static IWebHostBuilder UseStartup(this IWebHostBuilder hostBuilder, Func<WebHostBuilderContext, object> startupFactory)
+        public static IWebHostBuilder UseStartup<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]TStartup>(this IWebHostBuilder hostBuilder, Func<WebHostBuilderContext, TStartup> startupFactory)
```
Why isn't this `[DynamicallyAccessedMembers(StartupAccessibility)]`?

davidfowl replied:
It could be but you have to pass a factory with an instance so constructors are handled in this case.

eerhardt replied:
Oh, because the `Func` creates the Startup object.

eerhardt commented on src/Hosting/Hosting/src/Internal/StartupLinkerOptions.cs:
```diff
@@ -0,0 +1,15 @@
+using System;
```
(nit) copyright

davidfowl committed:
Fixed usings and copy right

eerhardt reviewed (approved):
These changes look good to me. Nice work.

Do you plan on having tests or analysis turned on in your build to ensure these don't get broken in the future?

davidfowl commented:
@eerhardt trying to decide if I should add a single test in our end to end tests that already publish bits instead of an individual tests.

eerhardt commented:
> trying to decide if I should add a single test in our end to end tests that already publish bits instead of an individual tests.

In a perfect world - I'd think both. But if you are choosing one, I'd choose end to end.

davidfowl committed:
Added a test to verify UseStartup and UseMiddleware works
- The linker.xml preserves constructors all of the types that are registered in DI. This should be removed once we get the linker friendly DI changes.

davidfowl commented:
OK added a test. It uses a Linker.xml file to keep constructors of types in DI. Most of this file should be gone once we get https://github.com/dotnet/runtime/pull/40227. The last thing looks like options https://github.com/dotnet/runtime/issues/40236.

davidfowl committed:
Skip OSX and Linux because we avoid restoring tests for other reasons

davidfowl committed:
Restore dependencies for linked application
- Add Restore option to DeploymentParameters

davidfowl committed:
Add retries

marek-safar commented:
@davidfowl it might be useful to enable linker warnings which would print if there are more cases to be fixed

eerhardt commented:
> it might be useful to enable linker warnings which would print if there are more cases to be fixed

For library authors (like ASP.NET here - and any library in dotnet/runtime that isn't part of the runtimepack) it would be useful to have an official linker option that runs the analysis/warnings on the whole library using the public surface area as the roots (like `-r assembly` does today).

That way library authors can easily tell which parts of their library aren't linker safe, and need to be fixed.

davidfowl commented:
I'll submit a follow up PR with any remaining warnings. I really want to try this with @eerhardt 's DI changes to see how close we are to no annotations.

Tratcher commented on src/Hosting/Server.IntegrationTesting/src/ApplicationPublisher.cs:
```diff
@@ -38,7 +38,7 @@ public virtual Task<PublishedApplication> Publish(DeploymentParameters deploymen
                                  + $" --no-dependencies"
                                  + $" /p:TargetArchitecture={deploymentParameters.RuntimeArchitecture}"
-                                 + " --no-restore";
+                                 + (deploymentParameters.RestoreDependencies ? "" : " --no-restore");
```
I recall no-restore is required for all tests on Helix because the test assets need to be pre-built before publishing to helix.

davidfowl replied:
Works fine on helix as far as I could tell.
